File: vscode-extension/src/mcpApproval.test.ts
import * as fs from "fs";
import * as os from "os";
import * as path from "path";

import { execFileSync } from "child_process";

import { checkMcpApproval, parseMcpGetOutput } from "./mcpApproval";

// Exact strings captured from a real `claude` CLI install -- see
// mcpApproval.ts's module docstring for how each was produced.
const REAL_PENDING_OUTPUT = `decoy:
  Scope: Project config (shared via .mcp.json)
  Status: ⏸ Pending approval (run \`claude\` to approve)
  Type: stdio
  Command: echo
  Args: hello
  Environment:
`;

const REAL_NOT_CONFIGURED_OUTPUT =
  'No MCP server named "nonexistent-server-xyz". .mcp.json servers are awaiting approval — run `claude` in this directory to review them.';

// The OTHER real "not found" message -- when no .mcp.json exists at all
// yet, rather than one existing with other unapproved entries. Both must
// classify as not_configured; confirmed these are worded differently
// enough that a marker matching only the first one would miss this case
// (verified directly: this exact gap failed once before the parser's
// marker was broadened from "awaiting approval" to "No MCP server named").
const REAL_NOT_CONFIGURED_NO_FILE_OUTPUT = 'No MCP server named "totally-unregistered-server". Run `claude mcp add` to add one.';

describe("parseMcpGetOutput (against real captured CLI output)", () => {
  it("detects Pending approval from a real `claude mcp get` response", () => {
    expect(parseMcpGetOutput(REAL_PENDING_OUTPUT, 0)).toBe("pending");
  });

  it("detects the not-configured case from a real error response (exit 1)", () => {
    expect(parseMcpGetOutput(REAL_NOT_CONFIGURED_OUTPUT, 1)).toBe("not_configured");
  });

  it("detects the OTHER real not-configured wording (no .mcp.json at all yet)", () => {
    expect(parseMcpGetOutput(REAL_NOT_CONFIGURED_NO_FILE_OUTPUT, 1)).toBe("not_configured");
  });

  it("treats an unrecognized non-zero exit as check_failed, not a guess", () => {
    expect(parseMcpGetOutput("some unexpected error we've never seen", 127)).toBe("check_failed");
  });

  it("treats exit 0 with neither marker as approved", () => {
    expect(parseMcpGetOutput("decoy:\n  Status: ✓ Connected\n", 0)).toBe("approved");
  });
});

describe("checkMcpApproval end-to-end against the REAL claude CLI", () => {
  const claudeAvailable = (() => {
    try {
      execFileSync("claude", ["--version"], { timeout: 5000 });
      return true;
    } catch {
      return false;
    }
  })();

  const maybeIt = claudeAvailable ? it : it.skip;

  maybeIt("reports 'pending' for a real, just-added, unapproved .mcp.json entry", async () => {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), "decoy-mcp-approval-test-"));
    try {
      execFileSync("git", ["init", "-q"], { cwd: dir });
      execFileSync("claude", ["mcp", "add", "--scope", "project", "decoy", "--", "echo", "hello"], {
        cwd: dir,
      });

      const result = await checkMcpApproval("decoy", dir);
      expect(result.status).toBe("pending");
      expect(result.raw).toContain("Pending approval");
    } finally {
      fs.rmSync(dir, { recursive: true, force: true });
    }
  }, 20000);

  maybeIt("reports 'not_configured' for a server name that was never registered", async () => {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), "decoy-mcp-approval-test-"));
    try {
      execFileSync("git", ["init", "-q"], { cwd: dir });
      const result = await checkMcpApproval("totally-unregistered-server", dir);
      expect(result.status).toBe("not_configured");
    } finally {
      fs.rmSync(dir, { recursive: true, force: true });
    }
  }, 20000);
});

File: intellij-plugin/core/src/test/kotlin/com/decoy/core/McpApprovalTest.kt
package com.decoy.core

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File

// Exact strings captured from a real `claude` CLI install -- see
// mcpApproval.ts's (VS Code side) module docstring for how each was
// produced; this Kotlin parser matches the same real output, not a
// separately-guessed IntelliJ-specific format (the `claude` CLI's output
// doesn't depend on which IDE is asking).
private const val REAL_PENDING_OUTPUT = """decoy:
  Scope: Project config (shared via .mcp.json)
  Status: ⏸ Pending approval (run `claude` to approve)
  Type: stdio
  Command: echo
  Args: hello
  Environment:
"""

private const val REAL_NOT_CONFIGURED_OUTPUT =
    "No MCP server named \"nonexistent-server-xyz\". .mcp.json servers are awaiting approval — run `claude` in this directory to review them."

private const val REAL_NOT_CONFIGURED_NO_FILE_OUTPUT =
    "No MCP server named \"totally-unregistered-server\". Run `claude mcp add` to add one."

class McpApprovalTest {

    @Test
    fun `parseMcpGetOutput detects Pending approval from real captured output`() {
        assertEquals(McpApprovalStatus.PENDING, parseMcpGetOutput(REAL_PENDING_OUTPUT, 0))
    }

    @Test
    fun `parseMcpGetOutput detects not-configured from real captured output (mcp json exists)`() {
        assertEquals(McpApprovalStatus.NOT_CONFIGURED, parseMcpGetOutput(REAL_NOT_CONFIGURED_OUTPUT, 1))
    }

    @Test
    fun `parseMcpGetOutput detects not-configured from real captured output (no mcp json at all)`() {
        assertEquals(McpApprovalStatus.NOT_CONFIGURED, parseMcpGetOutput(REAL_NOT_CONFIGURED_NO_FILE_OUTPUT, 1))
    }

    @Test
    fun `parseMcpGetOutput treats an unrecognized non-zero exit as CHECK_FAILED, not a guess`() {
        assertEquals(McpApprovalStatus.CHECK_FAILED, parseMcpGetOutput("some unexpected error", 127))
    }

    @Test
    fun `parseMcpGetOutput treats exit 0 with neither marker as APPROVED`() {
        assertEquals(McpApprovalStatus.APPROVED, parseMcpGetOutput("decoy:\n  Status: Connected\n", 0))
    }

    private fun claudeAvailable(): Boolean = try {
        ProcessBuilder("claude", "--version").start().waitFor() == 0
    } catch (_: Exception) {
        false
    }

    @Test
    fun `checkMcpApproval end-to-end reports PENDING for a real just-added unapproved entry`(@TempDir dir: File) {
        org.junit.jupiter.api.Assumptions.assumeTrue(claudeAvailable(), "claude CLI not available in this environment")
        ProcessBuilder("git", "init", "-q").directory(dir).start().waitFor()
        val add = ProcessBuilder("claude", "mcp", "add", "--scope", "project", "decoy", "--", "echo", "hello")
            .directory(dir).start()
        add.waitFor()

        val result = checkMcpApproval("decoy", dir)
        assertEquals(McpApprovalStatus.PENDING, result.status)
        assertTrue(result.raw.contains("Pending approval"))
    }

    @Test
    fun `checkMcpApproval end-to-end reports NOT_CONFIGURED for an unregistered name`(@TempDir dir: File) {
        org.junit.jupiter.api.Assumptions.assumeTrue(claudeAvailable(), "claude CLI not available in this environment")
        ProcessBuilder("git", "init", "-q").directory(dir).start().waitFor()

        val result = checkMcpApproval("totally-unregistered-server", dir)
        assertEquals(McpApprovalStatus.NOT_CONFIGURED, result.status)
    }
}
